rmgpy/data/kinetics/family.py

```diff
@@ -2698,149 +2809,332 @@ def extendNode(self,parent,thermoDatabase=None,obj=None,T=1000.0):
-                    del self.groups.entries[label]   
-                    del self.rules.entries[label]
-                    break
-                elif len(entry.parent.children) == 2: 
-                    child = [c for c in entry.parent.children if c != entry][0]
-                    if len(self.rules.entries[child.label]) == 0 and len(child.children) == 0:
-                        label = entry.parent.label
-                        entry.parent.parent.children.remove(entry.parent)
-                        entry.parent.parent.children.append(entry)
-                        entry.parent = entry.parent.parent
-                        del self.groups.entries[label]   
-                        del self.rules.entries[label]
-                        clabel = child.label
-                        del self.groups.entries[clabel]
-                        del self.rules.entries[clabel]
+                if entry.index != -1 and len(templateRxnMap[entry.label])>1 and entry not in multCompletedNodes:
+                    boo2 = self.extendNode(entry,templateRxnMap,thermoDatabase,obj,T)
+                    if boo2: #extended node so restart while loop
                         break
+                    else: #no extensions could be generated since all reactions were identical
+                        multCompletedNodes.append(entry)
             else:
                 boo = False
-            
+
             #fix indicies
             iters = 0
             for entry in self.groups.entries.itervalues():
                 if entry.index != -1:
                     entry.index = iters
                     iters += 1
-        
+
         return
-    
+
+
+    def makeBMRulesFromTemplateRxnMap(self, templateRxnMap):
+
+        index = max([e.index for e in self.rules.getEntries()] or [0]) + 1
+
+        for entry in self.groups.entries.values():
+            if entry.index == -1 or self.rules.entries[entry.label] != []:
+                continue
+            rxns = templateRxnMap[entry.label]
+            descendants = getAllDescendants(entry)
+            for entry2 in descendants:
+                rxns.extend(templateRxnMap[entry2.label])
+
+            assert rxns != [], entry.label
+            kinetics = ArrheniusBM().fitToReactions(rxns,family=self)
+
+            new_entry = Entry(
+                index = index,
+                label = entry.label,
+                item = self.forwardTemplate,
+                data = kinetics,
+                rank = 11,
+                reference=None,
+                shortDesc="BM rule fitted to {0} training reactions at node {1}".format(len(rxns),entry.label),
+                longDesc="BM rule fitted to {0} training reactions at node {1}".format(len(rxns),entry.label),
+            )
+            new_entry.data.comment = "BM rule fitted to {0} training reactions at node {1}".format(len(rxns),entry.label)
+
+            self.rules.entries[entry.label].append(new_entry)
+
+            index += 1
+
+
+
```

Comment:
Could you remove extra empty lines? Methods should be separated by 1 empty line.

Reply:
done in atg4